### DISCUSSION TITLE
How to send multiple streams on one PeerConnection ?

### DISCUSSION THREAD

**Discussion by acevest** [pion/webrtc#1955] [Q&A] [answered]
hi, i want to send multiple streams on one peerconnection.

i have tried this between webpage and webpage. it works.

![image](https://user-images.githubusercontent.com/7140471/133050178-10bd2a3d-0881-4832-8cdf-f70074cc3db0.png)

but when i tried `pion/webrtc` communicate with webpage, it seems not work.

```go
newVideoStream(iceConnectedCtx, peerConnection, videoFileName, "video_a", "pion_a")
newVideoStream(iceConnectedCtx, peerConnection, videoFileName, "video_b", "pion_b")
```

i can only got one stream `pion_a` on the webpage side，stream `pion_b` does not exist.

![image](https://user-images.githubusercontent.com/7140471/133055016-666f054a-3799-48fd-a942-408ac01370c5.png)

here is `newVideoStream`

```go
func newVideoStream(iceConnectedCtx context.Context, peerConnection *webrtc.PeerConnection, fileName, id, streamID string) {
	videoTrack, videoTrackErr := webrtc.NewTrackLocalStaticSample(webrtc.RTPCodecCapability{MimeType: webrtc.MimeTypeH264}, id, streamID)
	if videoTrackErr != nil {
		panic(videoTrackErr)
	}

	rtpSender, videoTrackErr := peerConnection.AddTrack(videoTrack)
	if videoTrackErr != nil {
		panic(videoTrackErr)
	}

	go func() {
		rtcpBuf := make([]byte, 1500)
		for {
			if _, _, rtcpErr := rtpSender.Read(rtcpBuf); rtcpErr != nil {
				return
			}
		}
	}()

	go func() {
		// Open a H264 file and start reading using our IVFReader
		file, h264Err := os.Open(fileName)
		if h264Err != nil {
			panic(h264Err)
		}

		h264, h264Err := h264reader.NewReader(file)
		if h264Err != nil {
			panic(h264Err)
		}

		// Wait for connection established
		<-iceConnectedCtx.Done()
		log.Printf("fucking video ...")
		ticker := time.NewTicker(h264FrameDuration)
		for ; true; <-ticker.C {
			nal, h264Err := h264.NextNAL()
			if h264Err == io.EOF {
				fmt.Printf("All video frames parsed and sent")
				os.Exit(0)
			}
			if h264Err != nil {
				panic(h264Err)
			}

			if h264Err = videoTrack.WriteSample(media.Sample{Data: nal.Data, Duration: time.Second}); h264Err != nil {
				panic(h264Err)
			}
		}
	}()
}
```

and init the peerconnection like this:

```go
	// Enable Extension Headers needed for Simulcast
	m := &webrtc.MediaEngine{}
	if err := m.RegisterDefaultCodecs(); err != nil {
		panic(err)
	}
	for _, extension := range []string{
		"urn:ietf:params:rtp-hdrext:sdes:mid",
		"urn:ietf:params:rtp-hdrext:sdes:rtp-stream-id",
		"urn:ietf:params:rtp-hdrext:sdes:repaired-rtp-stream-id",
	} {
		if err := m.RegisterHeaderExtension(webrtc.RTPHeaderExtensionCapability{URI: extension}, webrtc.RTPCodecTypeVideo); err != nil {
			panic(err)
		}
		if err := m.RegisterHeaderExtension(webrtc.RTPHeaderExtensionCapability{URI: extension}, webrtc.RTPCodecTypeAudio); err != nil {
			panic(err)
		}
	}

	// Create a InterceptorRegistry. This is the user configurable RTP/RTCP Pipeline.
	// This provides NACKs, RTCP Reports and other features. If you use `webrtc.NewPeerConnection`
	// this is enabled by default. If you are manually managing You MUST create a InterceptorRegistry
	// for each PeerConnection.
	i := &interceptor.Registry{}

	// Use the default set of Interceptors
	if err := webrtc.RegisterDefaultInterceptors(m, i); err != nil {
		panic(err)
	}

	config := webrtc.Configuration{
		ICEServers: []webrtc.ICEServer{
			{
				URLs: []string{"stun:stun.l.google.com:19302"},
			},
		},
	}

	// Create a new RTCPeerConnection
	//peerConnection, err := webrtc.NewPeerConnection(config)
	peerConnection, err := webrtc.NewAPI(webrtc.WithMediaEngine(m), webrtc.WithInterceptorRegistry(i)).NewPeerConnection(config)
	if err != nil {
		panic(err)
	}
```

so, my question is:
 1. does pion/webrtc support send multiple streams on one peerconnection ?
 2. if support, how to do it ?

**Comment by Sean-Der** [comment, marked as answer]
Hey @acevest 

pion/webrtc does support multiple audio/video tracks per PeerConnection.

I am not exactly sure what the problem is. I would try [play-from-disk-renegotation](https://github.com/pion/webrtc/tree/master/examples/play-from-disk-renegotation). This example shows adding/removing multiple tracks to a single PeerConnection.

**Reply by acevest** [reply]
thank you @Sean-Der 

with the help of this code [play-from-disk-renegotation](https://github.com/pion/webrtc/tree/master/examples/play-from-disk-renegotation) , i found a solution:

add this javascript code to the webpage

```javascript
pc.addTransceiver('video');
```

and then i got the stream `pion_b` on the webpage webrtc client side.

![image](https://user-images.githubusercontent.com/7140471/133222234-4caea52b-9a12-4db1-8f28-3783d5d49329.png)

but, it's a little different from 

![image](https://user-images.githubusercontent.com/7140471/133222305-dd3d90d9-c13c-4945-b96c-c2abdd9812d2.png)


`pion/webrtc` trigger `pc.ontrack` twice, each time the `event.streams.length == 1`

webrtc between webpage and webpage, trigger `pc.ontrack` once, with the `event.streams.length == 2`

do you know why ?

**Reply by Sean-Der** [reply]
When you are doing `webpage <-> webpage` you are doing a MediaStream with two MediaStreamTracks

When you are doing `pion <-> webpage` you are doing two MediaStreams each with one MediaStreamTrack.

You can re-create the `webpage <-> webpage` experience by creating two tracks with the same [streamId](https://github.com/pion/webrtc/blob/500338005497c222df5e1b3a2d5ac73a91cf07a5/track_local_static.go#L176)

**Reply by tuananhhust05** [reply]
Dear @acevest , i am new with webrtc, I'm working with a sfu project and have some issue with multiple stream, i sent multiple streams( exact 2) to server and restore it on array, after i tried send streams to an other client but this client just receive 1 stream. Can you give me some suggestions or keywords to resolve this problem ( receive mutiple stream with 1 peer connection) ,
 Thank you very much!
